### DISCUSSION TITLE
Node.js Binary Manager Summit?

### DISCUSSION THREAD

**Discussion by jasnell** [nodejs/help#4831] [Ideas]
@arcanis, @charlespierce, @dherman, @MylesBorins, @jasongin, @ljharb, @mhdawson, @nodejs/tsc and I'm sure many others...

As you all are aware, there has been discussion around adding corepack to the Node.js distribution in order to make it easier for yarn and pnpm users to use those tools out of the box with Node.js. Previously, we've had (unfortunately unsuccesful) discussions around the possibility of pulling version management tools like nvm and nvs into the Node.js project and distribution. The folks in the volta project (@charlespierce and @dherman) have reached out given the activity around corepack with an interest on discussing the long-term strategy for binary/version management ought to be.

What I'd like to do in the coming month or two (definitely before the end of the year) is organize a Binary Management Summit -- much in the same spirit we've had previously with things like the Diagnostics summit and other topics in the past, only virtual.

The key question for the summit would be this: If Node.js were to bundle a binary management utility as part of the core distribution, what would that look like? What would it's features be? What would it *not* do? How would we execute on it? For the sake of this discussion, "binary management utility" means a tool that can manage which versions of Node.js are installed on a system along with package management tools(e.g. npm, yarn, pnpm).

For the summit, what I propose are two four-hour working sessions split over two days in late November or early December. These would be working sessions with the specific goal of discussing and deciding a path forward, with specific actions identified to be taken to deliver a binary management utility sometime in the Node.js 16.x timeframe before Node.js 16 moves under LTS.

**Comment by MylesBorins** [comment]
I'm very excited about this idea!

I recently started experimenting with rust and have found `rustup` to be a fantastic tool that has enabled me to be super productive super quickly.

Have an official binary manager that we ship could significantly simplify a bunch of things imho. A big challenge here would be making sure we ship something that is self-reliant + multiplatform. It would appear that volta has an advantage here being written in rust, speaking only to cross-platform not API or capabilities.

We do need to be careful about timing around holidays. We should pick some days ASAP

**Comment by ljharb** [comment]
We already have a group to discuss this - https://github.com/nodejs/version-management - which volta is welcome to join. There hasn't been discussions in a long time because there are many things on which we'd all have to have common agreement, and each one is difficult to address.

**Reply by MylesBorins** [reply]
I think what James is suggesting is slightly different. My impression is that this is about holding a summit to talk about binary management and help to build consensus around what the problem space we want to solve is and the potential solutions. With corepack very close to being able to land in core this seems very time sensitive to ensure we have broad buy in on what path we want to take moving forward. The audience for this is broader than nodejs/version-management... as the question is about how we distribute node itself (e.g. do we want to ship something like rustup).

While we could use version-management as the venue for this discussion, I do think it needs to be a broader audience.

**Comment by bnb** [comment]
How should we mark interest in attendance/participation in this summit? Maybe just reply to this comment's thread with a +1 or note of interest in participation?

FWIW I'm extremely interested in seeing this happen.

**Reply by darcyclarke** [reply]
+1 would definitely like to be apart of this

**Comment by charlespierce** [comment]
Thanks for kicking this off @jasnell! We're super excited to see improvements in the developer ergonomics of installing Node. Having a clear delineation of the constraints and a vision of the end goal will make it a lot easier to take incremental steps towards that end vision.

**Comment by jasnell** [comment]
> I think what James is suggesting is slightly different. 

Yep. This isn't about setting up a team or a working group, it's about scheduling a couple of days to work out a concrete plan of action.

**Reply by ljharb** [reply]
Sounds great, looking forward to being involved.

**Comment by arcanis** [comment]
That sounds great! 🙂 Consider myself available mid November / early December, ideally between 7am - 19pm UTC.

**Comment by jasnell** [comment]
I'll start working on setting it up and will post a few options for days for folks to vote on and will start putting together an agenda.

**Reply by MylesBorins** [reply]
Would you consider making a doodle poll?

https://doodle.com/make-a-poll

**Comment by Trott** [comment]
+1, interested

**Comment by jasnell** [comment]
Ok, For everyone interested in participating, please pick two consecutive days from the choices here that work best: https://doodle.com/poll/z2dffebs6uz5q5da

I have the poll baselined for 9:30am Eastern or 10:30am Eastern. We can adjust start times as necessary depending on where folks are at.

**Reply by ljharb** [reply]
Unfortunately all of those times are much too early for me on the west coast; they conflict with my childcare responsibilities. Anything later in the day?

**Reply by MylesBorins** [reply]
We could hold more than a single session. This might make consensus harder... but finding times to accommodate NA / EMEA / APAC will be very tough

**Reply by jasnell** [reply]
I'd prefer not to split the group up. @ljharb, what's the earliest you can do?

**Reply by jasnell** [reply]
And... would actually shifting the meeting *earlier* be better. We could schedule in a break point for you to attend to morning family duty.

**Reply by ljharb** [reply]
Carving out a 4 hour swath is basically impossible for me in the mornings - every morning, I'm either with my kids, or have scheduled appointments. For example, I could do 8:30-9:30 AM PT on Monday and Wednesday mornings, or 8-10 AM PT on Friday mornings, but I only have a 4 hour block available during some workdays (or on the weekends). So, for me, later is better, if we can't go for shorter.

**Reply by MylesBorins** [reply]
@ljharb is there a way you could be flexible for a single day? Unfortunately not being able to start until after noon PST puts us past 6pm to start for most of Europe

**Reply by ljharb** [reply]
Could we do multiple 1-2 hour blocks instead of 4 hour blocks? It's much easier to rearrange things for shorter chunks of time.

**Reply by jasnell** [reply]
That might be difficult also because it makes scheduling other things a bit more difficult. Let's do this... let's focus on days for now, and once we have a good idea about what days may work, we'll look at the time blocking during those days.

**Reply by ljharb** [reply]
I mean, in that case I'm basically "maybe every day", since my available times vary widely from day to day.

**Comment by mhdawson** [comment]
I'm interested, filled in doodle.

**Comment by mhdawson** [comment]
Are are narrowing in on a date for the summit ?

**Comment by jasnell** [comment]
Yes, sorry, schedule got completely out of whack with nodeconf and other stuff. Returning to this now. Let me check the doodle but last I saw there really wasn't a great set of options. I'll summarize the options here.

**Comment by jasnell** [comment]
![image](https://user-images.githubusercontent.com/439929/98417930-bb710080-2036-11eb-9404-04a949bb9503.png)

Here's what we've got

**Reply by jasnell** [reply]
Looks like Nov 19 and 20 or Dec 3 and 4 are the days with the best likelihood of success

**Reply by ljharb** [reply]
11/16-11/19 is TC39 from 1AM - 6AM PT, so those mornings I'll be asleep. On other days, I will try to attend as much of whatever time is chosen as possible, but I unfortunately don't have any 4 hour blocks available in the mornings.

**Reply by jasnell** [reply]
Yeah, I get it. I have some conflicts in there also. This is going to be a best effort given the range of time zones we have here.

**Reply by bnb** [reply]
Would it be possible to put something on the Node.js Calendar for this?

**Reply by jasnell** [reply]
Yeah, let's go with Dec 3 and 4, 7:30am pacific (10:30am eastern) option. @bnb, I'm not sure if I have access to add it to the calendar, or if I do I've long lost the instructions on doing so. Could I ask you to take care of that?

**Reply by MylesBorins** [reply]
FYI the 4th conflicts with the OpenJS Board meeting which runs from 1pm ET - 3pm ET

**Comment by mhdawson** [comment]
@jasnell do you have info in terms of how to join etc.  I can add the calendar entry once I have the logistics to go along with it

**Reply by arcanis** [reply]
I'm fine with Dec 3/4, but note that I don't have access to the Node calendar.

**Reply by Trott** [reply]
> I'm fine with Dec 3/4, but note that I don't have access to the Node calendar.

I think they mean the Node.js calendar that is publicly viewable at https://nodejs.org/calendar.

**Reply by jasnell** [reply]
@mhdawson ... sorry I just saw the question here. I've set up a zoom link but will share the password separately like we do for TSC meetings. We'll just need folks to RSVP.

Dec 3 (7:30am - 11:30am Pacific) - https://nearform.zoom.us/j/96822233550
Dec 4 (7:30am - 11:30am Pacific) - https://nearform.zoom.us/j/95859553268

**Reply by charlespierce** [reply]
@jasnell I see the entry on the calendar with the Zoom links, but how do we RSVP to get the password?

**Reply by Trott** [reply]
How should folks RSVP? Will an email go out or something?

**Comment by mhdawson** [comment]
As mentioned by Myles the last 1 1/2 hours on the second day conflicts with the board meeting. Otherwise I'll plan to be there.

**Reply by mhdawson** [reply]
I've added to the public calendar, but we should add instructions in those entries on how to join

James can you add the text that should explain to people how to get the zoom/password?

**Reply by mhdawson** [reply]
I guess just how to get the password, added the zooms to the entries

**Comment by jasnell** [comment]
RSVP by dropping a comment in here. I'll send the password out to the email associated with your GitHub account. If you don't have an email associated or want me to use another, send me an email to let me know which you'd prefer.

**Reply by mhdawson** [reply]
Just in case this is my RSVP

**Reply by arcanis** [reply]
Will be there 👍

**Reply by ljharb** [reply]
I’ll be there for as much as i can

**Reply by ruyadorno** [reply]
+1

**Reply by charlespierce** [reply]
I'll be there :+1:

**Reply by MylesBorins** [reply]
I'll be there most of the time

**Reply by Trott** [reply]
RSVP'ing. I can't be there the whole time, but I would like to be there as much as I can.

**Reply by mmarchini** [reply]
Will try to join, not sure if I'll make it though

**Reply by jasnell** [reply]
You should all have received calendar invites.

**Reply by jasongin** [reply]
@jasnell I'll join also.

**Reply by jasnell** [reply]
@jasongin @darcyclarke ... I just sent you both the invite for the zoom call in email.

**Reply by jasnell** [reply]
The call is starting in about 5 minutes

**Comment by MylesBorins** [comment]
Do we have an agenda at all?

**Reply by jasnell** [reply]
I'll be sending out and update later today

**Comment by jasnell** [comment]
Agenda document here: https://docs.google.com/document/d/1Pqpw2Mua20uk6PjI24aXqHZZCe-Ux2KgrpaVqmDQVUM/edit?usp=sharing

Anyone with the link can view/comment. Editing will be opened at the start of the summit. We can use the same document for shared note taking during.

**Comment by mhdawson** [comment]
Did you send out an email with a password to join? I don't seem to be able to find it?

**Reply by VincentBailly** [reply]
I am also looking for the passcode.

**Reply by jasnell** [reply]
Link is in the Agenda doc now

**Comment by boneskull** [comment]
I’m not able to join but wanted to throw in an opinion:

Node **should** have an official tool _a la_ `rustup`, and should not punt on the question in the name of “encouraging competition.” 

If this means Node adopts a tool, or a new tool is written, or existing tools merge their efforts together (consolidation isn’t a dirty word!), so be it. An official tool will not prevent anyone from going off and writing their own. 

Would love to see something happen here even if it’s just adopting an existing tool. Thanks for working on this, whoever is involved!

**Reply by ruyadorno** [reply]
that's pretty much were we landed in the discussion 😊 https://github.com/nodejs/node/discussions/35682#discussioncomment-147436

we're going to have follow up calls to settle on _what_ tool that's going to be in case you're interested in joining us @boneskull

**Comment by jasnell** [comment]
FYI... I've updated the meeting join link for today's session in the top of the agenda document.

**Comment by jasnell** [comment]
FYI... we ended up ending a bit early today. We have some specific action items coming out of the call today... specifically:

1. I will be writing up a description of the hypothetical/fictional tool we discussed requirements for.
2. @charlespierce and @jasongin will be working on a summary/comparison of the existing tools in this space, building on the initial work that @jasongin started here: https://gistpreview.github.io/?cf9f64dd2739d78412bb9410701bf166/NodeVersionManagers.html
3. Starting in January 2021, we will organize regular bi-weekly calls to discuss and moving things forward. I will be setting up a doodle poll to begin getting dates / times for the next call.

On the call, we concluded with a consensus that there is a need for a tool that manages the node.js binary and associated toolchain. We agreed that a greenfield implementation of a brand new tool was impractical and not desirable and that the most likely course forward is adoption of one of the existing tools in this space. That *could* mean adopting it as is or forking it and iterating forward. 

The next time we get together, we will use the abstract writeup of the hypothetical tool and the survey of the existing tools to begin evaluating what our starting point is going to be.

tl;dr -- we want an officially blessed node.js version / toolchain manager that will be officially part of the node.js project (conceptually, think of it as something similar to rustup in the rust project) but we are still working on the specific details.

**Reply by arcanis** [reply]
Happy new year everyone! When should we hold the next call?

**Comment by designMoreWeb** [comment]
That sounds good @jasnell Looking forward to the next meeting/round table discussion

**Comment by mhdawson** [comment]
@jasnell thanks for leading and organising. I think it was a very productive and useful summit.

**Comment by ark120202** [comment]
Кто вы? ВАм нужен Аркадий? Если говорите по русски позвоните по телефону
